### ISSUE TITLE
bug: IndexError due to numeric precision logic

### ISSUE THREAD

**Issue by pnadolny13** [MeltanoLabs/target-snowflake#103]
Originally raised in slack https://meltano.slack.com/archives/C01TCRBBJD7/p1691415382778629

`IndexError: list index out of range`

The culprit looks to be this logic thats assuming a the numeric scale rules are floats when they should also support integers.

https://github.com/MeltanoLabs/target-snowflake/blob/045b614c19422e8ff2210207bb40a414f3c13e95/target_snowflake/connector.py#L242

**Comment by pnadolny13** [comment]
@edgarrmondragon I'm curious if you have thoughts on the best way to handle precision here. I reevaluated this logic and I'm now interpreting the json schema rules differently than before. I only see the multipleOf setting definitively affecting the precision. For example a minimum of 0.1 (precision 1) wouldn't restrict me from having a value of 0.15 (precision 2), so my assumption that the precision of the max/min schema value would define the precision of the column was wrong.

I'm inclined to default the precision to the maximum allowable of 37 decimal places unless multipleOf is set, in which case the precision would be the precision of the multiple of value i.e. multipleOf 0.1 would be precision 1. What do you think?

Obviously as part of this I'd fix the index bug thats causing this bug also.

**Comment by edgarrmondragon** [comment]
> I'm inclined to default the precision to the maximum allowable of 37 decimal places unless multipleOf is set, in which case the precision would be the precision of the multiple of value i.e. multipleOf 0.1 would be precision 1. What do you think?

@pnadolny13 That makes sense to me, it's the only relevant value to check to determine precision.

**Comment by pnadolny13** [comment]
@edgarrmondragon I'm second guessing that implementation now. I saw in the wise variant they use float for number json types https://github.com/transferwise/pipelinewise-target-snowflake/blob/aa6ac2e392b6148994f9138241630ba8d684adac/target_snowflake/db_sync.py#L90. I'm wondering if the whole thing would be solved more elegantly by just using the float type to support a loose definition for precision and scale within the snowflake table and defer the validation to the json schema validator in the target. If the record makes it through the json validator then it should meet expectations and the target should always accept it, there shouldnt ever really be a case where json validation passes but we need snowflake schemas to enforce types.

Also in snowflake allocating unused scale is wasteful unlike other type definition that dont affect the storage layer. https://docs.snowflake.com/en/sql-reference/data-types-numeric#impact-of-precision-and-scale-on-storage-size

Any thoughts?

cc @visch since I know you were dealing with floats in target-postgres at some point.

**Comment by visch** [comment]
Luckily for m on the postgres side I've been able to punt a lot as postgres has a concept of arbitrary [precision numbers](https://www.postgresql.org/docs/current/datatype-numeric.html#DATATYPE-NUMERIC-DECIMAL). I've messed with this a lot more with a custom target-mssql target I wrote. 

I think decimal can be handled with json schema fairly well actually, I'm going to save floats for the bottom of this.

I"m embarrassed about this code but it does encapsulate the issue well https://gitlab.com/autoidm/autoidm-target-mssql/-/blob/main/target_mssql/streams.py?ref_type=heads#L231-235 . I think taps should be required to define `json_maximum` and `json_multiple_of` which would allow us to calculate the proper Precision and Scale for the numeric type. 

Floats. I don't have an exact solution for but I think the appropriate thing to do for floats is in the tap side they should probably be defined as strings with a format of something like `single` `double` `quadruple` etc from https://en.wikipedia.org/wiki/Decimal32_floating-point_format . Luckily we haven't hit the float issue here so I think we should punt on that for now but I wanted to bring it up as it seems to come up everytime we talk about decimals, and I think this string implementation gives us the leeway to wait a bit
